include/lsst/meas/extensions/shapeHSM/HsmMomentsControl.h

```diff
@@ -1,46 +1,154 @@
 public:
-    HsmPsfMomentsControl() : ShapeControl("shape.hsm.psfMoments") {}
-    HsmPsfMomentsControl(std::string const & name) : algorithms::ShapeControl(name) {}
-    HsmPsfMomentsControl(HsmPsfMomentsControl const & other) : algorithms::ShapeControl(other) {}
+    /// A typedef to the Control object for this algorithm, defined above.
+    ///  The control object contains the configuration parameters for this algorithm.
+    typedef HsmMomentsControl Control;
+
+    /// @brief Initialize with standard field names and customized documentation.
+    HsmMomentsAlgorithm(Control const & ctrl, std::string const & name, afw::table::Schema & schema, char const* doc) :
+    _doc(doc),
+    _centroidResultKey(
+        base::CentroidResultKey::addFields(schema, name, "HSM Centroid", base::NO_UNCERTAINTY)),
+    _momentsKey(
+        base::ShapeResultKey::addFields(schema, name, "HSM moments", base::NO_UNCERTAINTY)
+    ),
+    _centroidExtractor(schema, name)
+    {}
+
+    /// Calculate moments
+    template<typename PixelT>
+    void calculate(
+        afw::table::SourceRecord& source, // Source for recording moments
+        PTR(afw::image::Image<PixelT>) const& afwImage, // Image on which to measure moments
+        PTR(afw::image::Mask<afw::image::MaskPixel>) const& afwMask, // Mask for image
+        afw::geom::Box2I const& bbox,     // Bounding box
+        afw::geom::Point2D const& center, // Starting center for measuring moments
+        afw::image::MaskPixel const badPixelMask, // Bitmask for bad pixels
+        float const width            // PSF width estimate, for starting moments
+        ) const;
+
+protected:
+    Control _ctrl;
+    std::string _doc;
+    base::CentroidResultKey _centroidResultKey;
+    base::ShapeResultKey _momentsKey;
+    base::FlagHandler _flagHandler;
+    base::SafeCentroidExtractor _centroidExtractor;
+};
+
+
+/// Class to measure HSM adaptive moments of source
+class HsmSourceMomentsAlgorithm : public HsmMomentsAlgorithm {
+public:
+    /// A typedef to the Control object for this algorithm, defined above.
+    ///  The control object contains the configuration parameters for this algorithm.
+    typedef HsmSourceMomentsControl Control;
+
+    /// @brief Initialize with standard field names and customized documentation.
+    HsmSourceMomentsAlgorithm(Control const & ctrl, std::string const & name, afw::table::Schema & schema) :
+        HsmMomentsAlgorithm(ctrl, name, schema, "Source adaptive moments algorithm from HSM"), _ctrl(ctrl) {}
+
 private:
-    virtual PTR(algorithms::AlgorithmControl) _clone() const;
-    virtual PTR(algorithms::Algorithm) _makeAlgorithm(
-        afw::table::Schema & schema, PTR(daf::base::PropertyList) const & metadata
+    Control _ctrl;
+    // These methods are defined private to prevent shadowing of the parent classes
+    void measure(
+        afw::table::SourceRecord & measRecord,
+        afw::image::Exposure<float> const & exposure
+    ) const;
```

Comment:
My assumption is that I'll wait until DM-980/1/2 all get merged, then update everything -- if that bumps my issue into S15, it won't worry me too much.